keps/sig-network/1672-tracking-terminating-endpoints/README.md

```diff
@@ -0,0 +1,152 @@
+terminating endpoints (see KEP-1669 as an example).
+
+## Motivation
+
+### Goals
+
+* Provide a mechanism to track whether an endpoint is terminating by only watching the EndpointSlice API.
+
+### Non-Goals
+
+* Consumption of the new API field is out of scope for this KEP but future KEPs will leverage
+the work done here to improve graceful terminination of pods in certain scenarios (see issue [85643](https://github.com/kubernetes/kubernetes/issues/85643))
+
+## Proposal
+
+This KEP proposes to keep "terminating" pods in the set of endpoints in EndpointSlice with
+additions to the API to indicate whether a given endpoint is terminating or not. If consumers
+of the API (e.g. kube-proxy) are required to treat terminating endpoints differently, they
+may do so by checking this condition.
+
+The criteria for a ready endpoint (pod phase + readiness probe) will not change based on the
+terminating state of pods, but consumers of the API may choose to prefer endpoints that are both ready and not terminating.
+
+### User Stories (optional)
+
+#### Story 1
+
+A consumer of the EndpointSlice API (e.g. kube-proxy) may want to know which endpoints are
+terminating without having to watch Pods directly for scalability reasons.
+
+One example would be the IPVS proxier which should set the weight of an endpoint to 0
+during termination and finally remove the real server when the endpoint is removed.
+Without knowing when a pod is done terminating, the IPVS proxy makes a best-effort guess
+at when the pod is terminated by looking at the connection tracking table.
+
+### Notes/Constraints/Caveats (optional)
+
+### Risks and Mitigations
+
+Tracking the terminating state of endpoints poses some scalability concerns as each
+terminating endpoint adds additional writes to the API. Today, a terminating pod
+results in 1 write in Endpoints (removing the endpoint). With the proposed changes,
+each terminating endpoint could result in at least 2 writes (ready -> terminating -> removed)
+and possibly more depending on how many times readiness changes during termination.
+
+## Design Details
+
+To track whether an endpoint is terminating, a `terminating` field would be added as part of
+the `EndpointCondition` type in the EndpointSlice API.
+
+```go
+// EndpointConditions represents the current condition of an endpoint.
+type EndpointConditions struct {
+    // ready indicates that this endpoint is prepared to receive traffic,
+    // according to whatever system is managing the endpoint. A nil value
+    // indicates an unknown state. In most cases consumers should interpret this
+    // unknown state as ready.
+    // +optional
+    Ready *bool `json:"ready,omitempty" protobuf:"bytes,1,name=ready"`
+
+    // terminating indicates if this endpoint is terminating. Consumers should assume a
+    // nil value indicates the endpoint  is not terminating.
+    // +optional
+    Terminating *bool `json:"terminating,omitempty" protobuf:"bytes,2,name=terminating"`
```

Comment:
Why a pointer here? Would it be safe to get rid of the nil case since it's indistinguishable from false here?

Reply:
I'm assuming that the nil state could be useful in the future. Some consumeres of EPSlice might have a special case for when the terminating state is unknown vs when it's explicitly false. Maybe I'm overthinking this?